### PR TITLE
[dy2static] support user to use decorator in their program

### PR DESCRIPTION
**Pull request by feifei-111**
### PR types

others
### PR changes

others
### Describe

1. user should only use 'to_static' to function 'forward'
2. call decorated functions in 'forward' is allowed
3. decorator will be transformed into function call
4. support decos with args
5. fixed a bug in return_transformer

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_decorator_transform.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+@paddle.jit.to_static
+def case2(x, y=0):
+    fun2(x, y)
+
+
+class TestDecoratorTransform(unittest.TestCase):
+
+    def test_deco_transform(self):
+        case1('case1', 5)
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_decorator_transform.py]: 按照单测规范，每一个test_XX里的测试内容都要有类似self.assert_XXX或者np.testing.assert_xxxx断言

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_decorator_transform.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import paddle.inference as paddle_infer
+from paddle.fluid.framework import program_guard, Program
+import numpy as np
+from paddle.fluid import core
+from functools import wraps
+
+
+def deco1(fun):
+
+    @wraps(fun)
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_decorator_transform.py]: 非wraps 的情况是否有异常捕获和对应的处理逻辑？

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        self.ancestor_nodes = []
+
+    def transform(self):
+        """
+        Main function to transform AST.
+        """
+        self.visit(self.root)
+
+    def visit(self, node):
+        self.ancestor_nodes.append(node)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py]: 移除 self.ancestor_node 无用的代码

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                    deco_name = deco.func.attr
+                else:
+                    deco_name = deco.func.id
+            else:
+                deco_name = deco.id
+            if deco_name in IGNORE_NAMES:
+                continue
+            decofun_str = '_jst.Call({})({})'.format(deco_name, decofun_str)
+
+        if decofun_str == '_orig_' + node.name:
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py]: 这里最好有一个flag状态量来判断

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    def visit_FunctionDef(self, node):
+        assert isinstance(node, gast.FunctionDef)
+        self.generic_visit(node)
+
+        deco_list = node.decorator_list
+        node.decorator_list = []
+
+        decofun_str = '_orig_' + node.name
+        for deco in reversed(deco_list):
+            if isinstance(deco, gast.Attribute):
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py:67]: 建议使用ast_to_source_code

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                deco_name = deco.attr
+            elif isinstance(deco, gast.Call):
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py:69]: 建议补充一个gast.Call的单测

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                continue
+            decofun_str = '_jst.Call({})({})'.format(deco_name, decofun_str)
+
+        if decofun_str == '_orig_' + node.name:
+            return node
+
+        orig_func_node = create_funcDef_node(node.body,
+                                             name='_orig_' + node.name,
+                                             input_args=node.args,
+                                             return_name_ids=[])
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py]: 可以看下如何优化下 create_funcDef_node函数逻辑，避免添加 return 空语句

File: python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+        decofun_str = '_decoed_{} = {}'.format(node.name, decofun_str)
+        decofun_node = gast.parse(decofun_str).body[0]
+
+        arg_str = ''
+        for arg in node.args.args:
+            arg_str += arg.id + ', '
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/decorator_transformer.py]: ",".join(ids)

**Review by 2742195759 — APPROVED** [review]
LGTM

**Pull request merged**